### PR TITLE
vm/sys_overlay Improvements

### PR DESCRIPTION
**Pull request by elad335**
- Implement sys_overlay_load_module_by_fd.
- Implement special segment allocation when ppc_seg flag is specified.

### CODE REVIEW

**Comment by kamer1337** [comment]
when i try to start Tony hawks proving ground and tony hawks project 8 i get this error

`F LDR: class std::runtime_error thrown: falloc() failed (addr=0x10000000, memsz=0x6caea)`

**Comment by elad335** [comment]
Log please.

**Comment by kamer1337** [comment]
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3296850/RPCS3.log.gz)

**Comment by elad335** [comment]
Should be working now.

**Comment by kamer1337** [comment]
same issue
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3296948/RPCS3.log.gz)

`·! 0:00:04.200779 LDR: ** Segment: p_type=0x1, p_vaddr=0x10000, p_filesz=0xdd7be8, p_memsz=0xdd7be8, flags=0x5
·! 0:00:04.242326 LDR: ** Segment: p_type=0x1, p_vaddr=0xdf0000, p_filesz=0x98f30, p_memsz=0x98f30, flags=0x6
·! 0:00:04.243700 LDR: ** Segment: p_type=0x1, p_vaddr=0x10000000, p_filesz=0x6caea, p_memsz=0x6caea, flags=0x400004
·! 0:00:04.244765 LDR: ** Segment: p_type=0x1, p_vaddr=0x10070000, p_filesz=0x172100, p_memsz=0x7b61d8, flags=0x600006
·F 0:00:04.247430 LDR: class std::runtime_error thrown: falloc() failed (addr=0x10070000, memsz=0x7b61d8)
(in file c:\projects\rpcs3\rpcs3\emu\cell\ppumodule.cpp:1092)`

**Comment by elad335** [comment]
Retest.

**Comment by kamer1337** [comment]
its good now

**Comment by jenci8888** [comment]
I've retested your PR, I found problems:
GTHD Concept (JP)
`F LDR: class std::runtime_error thrown: falloc() failed (addr=0x10000000, memsz=0x47800)`
GT5 Prologue (JP)
`F LDR: class std::runtime_error thrown: falloc() failed (addr=0x10000000, memsz=0x83780)`
NFS Pro Street
`F LDR: class std::runtime_error thrown: falloc() failed (addr=0x10000000, memsz=0xd0b00)`
NFS Carbon
`F LDR: class std::runtime_error thrown: falloc() failed (addr=0x10000000, memsz=0x40954)`
GTHD Concept (EU)
`F LDR: class std::runtime_error thrown: falloc() failed (addr=0x10000000, memsz=0x3ed80)`

**Comment by AniLeo** [comment]
Should be good, no review against

**Comment by illusion98** [comment]
Looks good 👍 
Let's wait for Neko to review/merge and we're good to go!

**Comment by elad335** [comment]
I would really like to hear whst @Nekotekina thinks about it, if he requests to dig into the kernel much further I can.

**Comment by L1ttl3c0w** [comment]
@Nekotekina Could you please review this?

**Comment by elad335** [comment]
Due to new finding of ppc seg segment must be alocated at fixed 0x30000000 from sys_process_spawn reversing ([source](https://user-images.githubusercontent.com/18193363/62004132-c6499f80-b129-11e9-98d1-c08e34e5e50e.png)) and making this pull request extremely fast to review/merge, I've removed .rsx_image, vm::main, and vm::user64k segments changes.

File: rpcs3/Emu/Cell/lv2/sys_overlay.cpp [hunk 1]
```diff
@@ -41,6 +38,49 @@ error_code sys_overlay_load_module(vm::ptr<u32> ovlmid, vm::cptr<char> path2, u6
...
+	}
+
+	const auto file = idm::get<lv2_fs_object, lv2_file>(fd);
+
+	if (!file)
+	{
+		return CELL_EBADF;
+	}
+
+	return overlay_load_module(ovlmid, fmt::format("%s_x%x", file->name.data(), offset), flags, entry);
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/lv2/sys_overlay.cpp]: It won't work like this, it needs to create a file view (lv2_file::make_view) from specified offset.

**Pull request merged**